lib/terminus.js

```diff
@@ -1,134 +1,148 @@
+    } catch (error) {
+      listener(req, res)
+    }
+
+    res.setHeader('Content-Type', 'application/json')
+
+    if (statusResponse.status === 'ok') {
+      res.statusCode = 200
+    } else {
+      res.statusCode = 503
+    }
 
-function decorateWithSignalHandler (server, state, options) {
-  const { signals, onSignal, beforeShutdown, onShutdown, timeout, logger } = options
+    return res.end(JSON.stringify(statusResponse))
+  }
+
+  _decorateWithHealthCheck () {
+    this.server.listeners('request')
+      .forEach((listener) => {
+        this.server.removeListener('request', listener)
+        this.server.on('request', async (req, res) => this._onRequest(req, res, listener))
+      })
+  }
 
-  stoppable(server, timeout)
+  _decorateWithSignalHandler () {
+    stoppable(this.server, this.options.timeout)
 
-  const asyncServerStop = promisify(server.stop).bind(server)
+    this._asyncServerStop = promisify(this.server.stop).bind(this.server)
+    this.options.signals.forEach(
+      sig => {
+        process.on(sig, () => this._cleanup())
+      }
+    )
+  }
 
-  async function cleanup (signal) {
-    if (!state.isShuttingDown) {
-      state.isShuttingDown = true
+  getHttpServer () {
+    return this.server
+  }
+  async getHealthStatus (url) {
+    if (this.options.healthChecks[url]) {
+      if (this.state.isShuttingDown) {
+        return this._sendFailure()
+      }
+      let info
       try {
-        await beforeShutdown()
-        await asyncServerStop()
-        await onSignal()
-        await onShutdown()
-        signals.forEach(sig => process.removeListener(sig, cleanup))
-        process.kill(process.pid, signal)
+        info = await this._executeHealtCheck(url)
       } catch (error) {
-        logger('error happened during shutdown', error)
-        process.exit(1)
+        this.options.logger('healthcheck failed', error)
+        return this._sendFailure(error.causes)
       }
+      return this._sendSuccess(info)
+    } else {
+      throw new HealthCheckNotFoundError('Given health check is not registered')
     }
   }
-  signals.forEach(
-    sig => process.on(sig, cleanup)
-  )
 }
 
-function terminus (server, options = {}) {
-  const { signal = 'SIGTERM',
-    signals = [],
-    timeout = 1000,
-    healthChecks = {},
-    onShutdown = noopResolves,
-    beforeShutdown = noopResolves,
-    logger = noop } = options
-  const onSignal = options.onSignal || options.onSigterm || noopResolves
-  const state = Object.assign({}, intialState)
-
-  if (Object.keys(healthChecks).length > 0) {
-    decorateWithHealthCheck(server, state, {
-      healthChecks,
-      logger
-    })
-  }
-
-  // push the signal into the array
-  // for backwards compatability
-  if (!signals.includes(signal)) signals.push(signal)
-  decorateWithSignalHandler(server, state, {
-    signals,
-    onSignal,
-    beforeShutdown,
-    onShutdown,
-    timeout,
-    logger
-  })
-
-  return server
+function createTerminus (server, options = {}) {
```

Comment:
@gergelyke as you can see `createTerminus` is unneeded. I left it in, so it would not break too much, but I can remove it and expose the Terminus class instead if you approve this.